src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -601,94 +590,116 @@ public static (RegexNode LoopNode, (char Char, string? String, char[]? Chars) Li
-                            concatNode = curNode;
-                            nextChild = 0;
+                            int childCount = node.ChildCount();
+                            RegexNode? child = null;
+                            if (leading)
+                            {
+                                for (int i = 0; i < childCount; i++)
+                                {
+                                    if (node.Child(i).Kind is not (RegexNodeKind.Empty or RegexNodeKind.PositiveLookaround or RegexNodeKind.NegativeLookaround))
+                                    {
+                                        child = node.Child(i);
+                                        break;
+                                    }
+                                }
+                            }
+                            else
+                            {
+                                for (int i = childCount - 1; i >= 0; i--)
+                                {
+                                    if (node.Child(i).Kind is not (RegexNodeKind.Empty or RegexNodeKind.PositiveLookaround or RegexNodeKind.NegativeLookaround))
+                                    {
+                                        child = node.Child(i);
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (child is not null)
+                            {
+                                node = child;
+                                continue;
+                            }
+
+                            goto default;
                         }
-                        break;
 
-                    case RegexNodeKind.Atomic:
-                    case RegexNodeKind.Capture:
-                        curNode = curNode.Child(0);
-                        concatNode = null;
-                        continue;
+                    case RegexNodeKind.Alternate:
+                        // For alternations, every branch needs to lead or trail with the same anchor.
+                        {
+                            // Get the leading/trailing anchor of the first branch.  If there isn't one, bail.
+                            RegexNodeKind anchor = FindLeadingOrTrailingAnchor(node.Child(0), leading);
+                            if (anchor == RegexNodeKind.Unknown)
+                            {
+                                return RegexNodeKind.Unknown;
+                            }
 
-                    case RegexNodeKind.Empty:
-                    case RegexNodeKind.PositiveLookaround:
-                    case RegexNodeKind.NegativeLookaround:
-                        break;
+                            // Look at each subsequent branch and validate it has the same leading or trailing
+                            // anchor.  If any doesn't, bail.
+                            int childCount = node.ChildCount();
```

Comment:
NIT: since we are visiting every child here, would it be worth to check if each child's lengths is the same, just so that if that's the case we return FindOptimization.TrailingAnchor_FixedLength?

Reply:
I'm not following. Can you elaborate?

Reply:
Sure, I'm actually rethinking this and I believe that the scenario I was thinking doesn't really apply and doesn't make sense, but I'll elaborate in what I was thinking anyway just in case. I was thinking of a case where you have a pattern like `((abc$|def$)+)`. In this case, we would compute min length of the node to be 0, and Max length to be 3, so the optimization you made for TrailingAnchor_FixedLength wouldn't be applied. That said, I was wondering if it was valuable to still analyze that for the alternation of the pattern, all branches have the same Length, so I was wondering if we could somehow make this be a sub-case of FixedLength optimization, since if the input is not empty, we can immediately navigate to (input.Length - 3) position in FindFirstChar

Reply:
> where you have a pattern like `((abc$|def$)+)`. In this case, we would compute min length of the node to be 0, and Max length to be 3

FWIW, we'd actually compute the min to be 3 and the max to be null (unbounded).  It's still the case as you say that TrailingAnchor_FixedLength wouldn't kick in.

It's not clear to me how to reason about an example like you cited.  An end anchor inside of a loop is strange (unless RegexOptions.Multiline is used, in which case it becomes an end-of-line anchor), since an iteration matching means subsequent iterations will fail.  We could thus consider automatically changing `(abc$|def$)+` to just `(abc$|def$)`, since we'll never match more than a single iteration, and then the fixed length optimization would actually kick in, but... this is one of those cases where I don't think it's worth us optimizing in patterns that don't make a lot of sense.  Unless I'm misreading it or misunderstanding it?

Reply:
I wrote the incorrect loop operator, I meant `(abc$|def$)?` which is why I was saying the min would be 0 and max 3.

Reply:
Ah, ok :smile: